Create Clavister.pm

Henningen opened the pull request:
I personally combine this with below yaml config under device_auth and use ssh keys for authentication.
Both ssh key, a user with auditor rights, and snmp (for inital discovery) needs to be enabled on the Clavister device in addition.
  - tag: 'clavister-walls'
    driver: cli
    platform: Clavister
    username: netdisco
    only:
      - 'description:Clavister.*'

inphobia commented:
i have a few minor remarks but this seems good.

the bigger question i have is how do you figure out what device to use it on, since i don't think snmp::info has any support for this? i had a look at their website and demo site, and they do have the option to enable snmp, but i found no mib or even a reference to a mib. using the description field is actually a pretty nifty option, but another option could be to use a speudodevice.

i guess i'm mostly wondering what the info is you get from an snmpwalk starting at .1.3.6.1.2.1.1 ? also, if you know they have a mib for their device that would help a lot too. even the most basic support in snmp::info would make things clearer and easier for other users.

thx

Henningen commented:
A few notes from me as well:

Currently, for the sshcollector I'm using the neighborcache command, with sample output like this (I removed wan devices ip):
```
kamos-fw:/> neighborcache 
Contents of Active Neighbor Cache [Active:13]
Idx Iface          HW Address        IP4 Address     State    Timeout Vendor
--- -------------- ----------------- --------------- -------- ------- ------
1   LAN            00-0C-29-6B-7D-B6 10.40.0.19      Active   0       VMware Inc.
2   LAN            A0-8C-FD-28-FC-52 10.40.0.95      Active   0       Hewlett-Packard Company
3   LAN            60-02-92-1E-66-4A 10.40.0.91      Active   0       Pegatron Corporation
4   WAN            7C-69-F6-7D-70-E0 X.X.X.X         Active   0       Cisco Systems Inc.
5   LAN            00-0C-29-00-A7-7D 10.40.0.21      Active   0       VMware Inc.
6   LAN            00-0C-29-58-14-D0 10.40.0.22      Active   0       VMware Inc.
7   LAN            8C-DC-D4-7F-F8-45 10.40.0.83      Active   0       Hewlett-Packard Company
8   LAN            00-0F-FE-9A-A1-3F 10.40.0.92      Active   0       G-PRO Computer
9   LAN            00-20-6B-96-EC-BB 10.40.0.203     Active   0       Konica Minolta Holdings Inc.
10  LAN            74-44-01-49-B2-AF 10.40.0.93      Active   0       Netgear
11  LAN            70-54-D2-7D-90-15 10.40.0.106     Active   0       Pegatron Corporation
12  LAN            8C-DC-D4-7F-FA-B0 10.40.0.80      Active   0       Hewlett-Packard Company
13  LAN            C8-D9-D2-82-32-AC 10.40.0.82      Inactive 160287  Hewlett-Packard Company
14  LAN            A0-8C-FD-28-FE-95 10.40.0.94      Inactive 105099  Hewlett-Packard Company
15  LAN            28-80-23-6E-0E-80 10.40.0.81      Active   0       Hewlett-Packard Company
```

It might be better/more sane to use the arp command on the devices, I'm not entirely sure if it will make much of a difference, I pretty much hacked this togheter with neighbourcache just because that was what I have been using to check the tables.

Arp command looks like this (I removed wan device ip):

```
kamos-fw:/> arp
ARP cache of iface WAN
  Dynamic X.X.X.X          = 7C-69-F6-7D-70-E0   Expire=576

ARP cache of iface LAN
  Dynamic 10.40.0.19       = 00-0C-29-6B-7D-B6   Expire=630
  Dynamic 10.40.0.21       = 00-0C-29-00-A7-7D   Expire=586
  Dynamic 10.40.0.22       = 00-0C-29-58-14-D0   Expire=576
  Dynamic 10.40.0.106      = 70-54-D2-7D-90-15   Expire=894
  Dynamic 10.40.0.91       = 60-02-92-1E-66-4A   Expire=613
  Dynamic 10.40.0.92       = 00-0F-FE-9A-A1-3F   Expire=702
  Dynamic 10.40.0.93       = 74-44-01-49-B2-AF   Expire=809
  Dynamic 10.40.0.95       = A0-8C-FD-28-FC-52   Expire=602
  Dynamic 10.40.0.80       = 8C-DC-D4-7F-FA-B0   Expire=612
  Dynamic 10.40.0.81       = 28-80-23-6E-0E-80   Expire=305
  Dynamic 10.40.0.83       = 8C-DC-D4-7F-F8-45   Expire=630
  Dynamic 10.40.0.203      = 00-20-6B-96-EC-BB   Expire=699
```

Here's a quick dump of the snmpwalk you requested:

```
# snmpwalk -v2c -cpublic 10.40.0.1 .1.3.6.1.2.1.1
iso.3.6.1.2.1.1.1.0 = STRING: "Clavister cOS Core 12.00.21.01-34783 "
iso.3.6.1.2.1.1.2.0 = OID: iso.3.6.1.4.1.5089.1
iso.3.6.1.2.1.1.3.0 = Timeticks: (143298700) 16 days, 14:03:07.00
iso.3.6.1.2.1.1.4.0 = STRING: "Serit iDrift"
iso.3.6.1.2.1.1.5.0 = STRING: "kamos-fw"
iso.3.6.1.2.1.1.6.0 = STRING: "serverrom"
iso.3.6.1.2.1.1.7.0 = INTEGER: 4
```



They seem to be detected as layer 3 devices and find interfaces and other info through generic SNMP as far as I can tell.

I've attached a few (lighty censored, sorry) screenshots of a sample device (Please note that link discovery is not working, I added the link manually in the web-gui):
![claviseter_sample  - Details](https://user-images.githubusercontent.com/12282900/67943788-204f7a80-fbdb-11e9-97ba-1594a9d4f1c2.png)
![clavister_sample   - Ports](https://user-images.githubusercontent.com/12282900/67943792-2180a780-fbdb-11e9-8696-9c9a33166562.png)
![clavister_sample - Modules](https://user-images.githubusercontent.com/12282900/67943799-27768880-fbdb-11e9-9bc3-485e88b4e41c.png)


The devices are running a custom built small fingerprint OS, and they don't support LLDP/CDP or simular discovery protocols currently (we do have a request for enhancement registered with them to maybe get that in the future, we should probably ask for mac-table as well).

The actual MIBS for clavister, seem to be for more "specialized functionality", but I'm not very fluent in SNMP and parsing so maybe I missed something. I concluded after some experiements that I was not able to get the ARP/MAC tables from there, but I might be giving up early.

The actual .mib(s) are downloaded from the web-gui on the devices/firewalls, under the path Status->Maintenance->Resources . I've attached some I downloaded from an Clavister E80 at our location.

Screenshot showing where:
![clavister_download_mibs](https://user-images.githubusercontent.com/12282900/67943039-7fac8b00-fbd9-11e9-8dff-2c842a261849.png)

And a copy of the MIBS I was able to download in gz format
[CLAVISTER-SMI.mib.gz](https://github.com/netdisco/netdisco/files/3793505/CLAVISTER-SMI.mib.gz)
[CLAVISTER-MIB.mib.gz](https://github.com/netdisco/netdisco/files/3793506/CLAVISTER-MIB.mib.gz)
[CLAVISTER-TRAPS-MIB.mib.gz](https://github.com/netdisco/netdisco/files/3793507/CLAVISTER-TRAPS-MIB.mib.gz)


Hope this answered some of your questions, feel free to ask more if required.

inphobia commented:
cool, i've just imported the mibs. ( https://github.com/netdisco/netdisco-mibs/pull/111 )

imo the easiest option would be to make a basic snmp::info module, this way netdisco will know clavister as a vendor and perhaps we can do some regex work to get basic info like os version etc... working. does the device show duplex/speed info with the generic layer3 module? since it seems those things are available in the mibs. custom names also seem to be included there.

the most basic snmp::info module should only take a few days, but an snmprec data dump (https://github.com/netdisco/snmp-info/wiki/Simulating-Agents#22-snmpsim--snmprec-version) would help if that's an option for you. do read the privacy reminders a bit higher on the same page. snmpwalk with data obfuscation is an option, but snmprec is prefered due to no data mangling. we/i will not share your data with anyone, not even other devs unless you give the ok. this will also let us figure out if mac and/or arp tables could be fetched via snmp.

you can get a hold of me either on irc: https://webchat.freenode.net/?randomnick=1&prompt=1&channels=%23netdisco
of via email: nick . nauwelaerts /at/ aquafin be

inphobia commented:
side note: do we want to include neighborcache entries labeled "Inactive"?
